### PR TITLE
Use html.standard rules for when a semicolon is optional after `&name`

### LINKED ISSUE #193: decoded html entities in href
href attribute value has decoded html entities in output.

Example
```java
        String input = "" +
                "<a href=\"/index.php?action=1&order_id=1\">order</a>";
        System.out.println(Sanitizers.LINKS.sanitize(input));
```

output
```
<a href="/index.php?action&#61;1ℴ_id&#61;1" rel="nofollow">order</a>
```

`&order` is decoded as `ℴ`, which renders invalid link

### PR DESCRIPTION
**Pull request by mikesamuel**
Fixed #193: &order in link broken

### CODE REVIEW

File: src/main/java/org/owasp/html/HtmlEntities.java [hunk 1]
```diff
@@ -2349,6 +2454,13 @@ public static int appendDecodedEntity(
     }
   }
 
+  private static boolean isHtmlIdContinueChar(char ch) {
```
Comment (yangbongsoo) [inline src/main/java/org/owasp/html/HtmlEntities.java:2457]: @mikesamuel 
What is the role of this method?

for example in this below case,
```java
assertEquals(
            "&#\n",
            Encoding.decodeHtml("&#&#xa;"));
```
I don't know how it depends on the result of isHtmlIdContinueChar.

```java
if (!isHtmlIdContinueChar(digit)) {
    if (i == offset + 2) {
        codepoint = -1;
    }
    if (digit == ';') {
        i += 1;
    }
    tail = i;
    break;
}
```

**Review by yangbongsoo — COMMENTED** [review]
I asked some questions because the behavior of the htmlEntities decode is complicated.

And I asked you about your thoughts on contribute for sanitizer by email, so please give me your opinion. thank you.

File: src/main/java/org/owasp/html/HtmlEntities.java (hunk 1 above)
Comment (yangbongsoo) [reply src/main/java/org/owasp/html/HtmlEntities.java:2457]: I find it. in below case, `&#x6a` check is needed. I got it.
```java
String attackString = "<a href=\"&#x6a;avascript:alert(1)\">XSS</a>";
String s = Encoding.decodeHtml(attackString);
```

File: src/main/java/org/owasp/html/HtmlEntities.java [hunk 2]
```diff
@@ -2320,23 +2416,32 @@ public static int appendDecodedEntity(
...
         char nameChar = html.charAt(i);
         t = t.lookup(nameChar);
         if (t == null) { break; }
+        if (t.isTerminal()) {
+          longestDecode = t;
+          tail = i + 1;
+        }
       }
-      if (t == null) {
+      if (longestDecode == null) {
```
[the remark this answers is not in the document]
Comment (mikesamuel) [reply src/main/java/org/owasp/html/HtmlEntities.java:2430]: longestDecode will be null if we never see a trie element where `t.isTerminal()`.

Imagine we have a Trie built from just the strings "\&Dot;" and "\&DotDot;"

    'D'═'o'═'t'╦'D'═'o'═'t'═';'
               ║
               ╚';'

In this case we have only two terminal nodes: the two ';' nodes.
Looking for entities in " \&DotD" we would traverse several Trie nodes but without finding a terminal node.

**Comment by mikesamuel** [comment]
Sorry for the delay in responding to your question.  It dropped off my GH notifications somehow.

File: src/main/java/org/owasp/html/HtmlEntities.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (yangbongsoo) [reply src/main/java/org/owasp/html/HtmlEntities.java:2430]: thank you for your detail explain :)

**Pull request merged**